README.md
# Mamba

Mamba is a framework to write, compile, and deploy smart contracts
written in Vyper language and Solidity language. On top of that, it has supports for writing
and testing decentralized applications using Web3.py and Pytest.

## Dependencies

* [Ganache](https://www.trufflesuite.com/ganache)
* [Ganache CLI](https://github.com/trufflesuite/ganache-cli)
* [Pip](https://pypi.org/project/pip/)
* [Python3](https://www.python.org/downloads/) version 3.6 or greater
* [Go Ethereum](https://geth.ethereum.org/downloads/)
* [Vyper](https://github.com/ethereum/vyper)
* [Solidity](https://soliditylang.org)

## Installation

You can install the latest release via ``pip``:

```bash
$ pip install black-mamba
```

To use Solidity, you must install ``solc`` compiler separately.

## Quick Usage

To set up the the structure of Mamba project directory:

```bash
$ mamba init
$ edit contracts/HelloWorld.vy
$ mamba compile
```

## Documentation

Mamba documentation can be found at [Mamba website](https://mamba.black/documentation).


## Contributing

I invite you to join Mamba squad! You can contribute to Mamba by
writing documentation, finding bugs, and creating pull requests.


docs/releases.rst
Release Notes
=============

v0.6.4 (2021-05-16)
-------------------

Bug Fixes
~~~~~~~~~

- Fix bug when testing Solidity contract

v0.6.3 (2021-05-16)
-------------------

Bug Fixes
~~~~~~~~~

- Fix bug when testing Solidity contract

v0.6.2 (2021-05-06)
-------------------

Features
~~~~~~~~

- Update Vyper, Web3.py, PyEVM, eth-tester version
- Use new EthereumTesterProvider for every test method

v0.6.1 (2021-02-17)
-------------------

Features
~~~~~~~~

- Add Solidity support for deploying smart contracts

v0.6 (2021-02-16)
-----------------

Features
~~~~~~~~

- Add Solidity support

v0.5.8 (2021-02-03)
-------------------

Features
~~~~~~~~

- Move methods in tokens.py file to accounts.py file

v0.5.7 (2021-02-03)
-------------------

Features
~~~~~~~~

- Add Accounts and Tokens methods for Etherscan

v0.5.6 (2021-01-20)
-------------------

Updates
~~~~~~~

- Update vyper to 0.2.8
- Update pytest to 6.2.1
- Update web3 to 5.15.0
- Update pytest-mock to 3.5.1
- Update eth-tester to 0.5.0b3

v0.5.5 (2020-10-23)
-------------------

Updates
~~~~~~~

- Update py-evm to 0.3.0a20
- Remove eth-tester as dependency for now

v0.5.4 (2020-10-22)
-------------------

Updates
~~~~~~~

- Update Vyper to 0.2.7
- Put py-evm on top for dependencies issue

v0.5.2 (2020-09-25)
-------------------

Bug Fixes
~~~~~~~~~

- Fix EthereumTesterProvider bug when accessing the balance of the accounts
- Add flag for Mamba CLI to execute test with Pytest

v0.5.1 (2020-09-24)
-------------------

Features
~~~~~~~~

- Update Vyper to 0.2.4

v0.5 (2020-09-23)
-----------------

Features
~~~~~~~~

- Add support for Ganache CLI for testing

v0.4.1 (2020-05-25)
-------------------

Features
~~~~~~~~

- Add method to pin manifest file

v0.4.0 (2020-05-18)
-------------------

Features
~~~~~~~~

- Add method to install, list, delete Ethereum packages
- Add method to create a manifest file

v0.3.5 (2020-05-08)
-------------------

Features
~~~~~~~~

- Use web3.py version 5.9.0

v0.3.4 (2020-05-01)
-------------------

Features
~~~~~~~~

- Add support for encrypted private key file
- Add Docker support

v0.3.3 (2020-04-16)
-------------------

Features
~~~~~~~~

- Add support for Infura


black_mamba/mamba_cli.py
from argparse import ArgumentParser
from pathlib import Path
from subprocess import run

from black_mamba.initialization.init import initialize_project_directory
from black_mamba.compilation.vyper_compiler import compile_all_files
from black_mamba.server.geth_dev_server import run_server
from black_mamba.auth import Authentication
from black_mamba.mnemonic import Mnemonic
from black_mamba.epm.package_manager import PackageManager
from black_mamba.constants import ETHEREUM_PACKAGES_DIR


def parse_cli_args():

    parser = ArgumentParser(description="Mamba Framework")
    parser.add_argument('mode', type=str, help="Mode of Mamba CLI: init / compile / server / keyfile")
    parser.add_argument('--httpserver', action="store_true", default=False)
    parser.add_argument('--websocket', action="store_true", default=False)

    keyfile_group = parser.add_argument_group('keyfile')
    keyfile_group.add_argument('--keyfile_mode', action="store", help="encrypt or decrypt the keyfile", default=False)
    keyfile_group.add_argument('--keyfile_file', action="store", default=False)
    keyfile_group.add_argument('--keyfile_private_key', action="store", default=False)
    keyfile_group.add_argument('--keyfile_password', action="store", default=False)

    epm_group = parser.add_argument_group('epm')
    epm_help = """Mode of Ethereum package manager: install / list / uninstall / create.
    "install" -> Install package from GitHub or registry.
    "list" -> List all installed packages.
    "uninstall" -> Uninstall package.
    "create" -> Create a package manifest from the source code, ABI, and bytecode of the smart contract.
    "pin" -> Pin the package manifest to IPFS.
    """
    epm_group.add_argument('--epm_mode', action="store", help=epm_help, default="list")
    epm_group.add_argument('--epm_uri', action="store", default=None)
    epm_group.add_argument('--epm_package', action="store", default=None)

    mnemonic_group = parser.add_argument_group('generate_mnemonic')
    mnemonic_group.add_argument('--num_words', action="store", help="number of words in mnemonic (must be 12, 15, 18, 21, 24)", default=12)
    mnemonic_group.add_argument('--language', action="store", help="language to use (must be english, czech, french, italian, japanese, korean, spanish, chinese_simplified, chinese_traditional)", default="english")

    test_group = parser.add_argument_group('test')
    test_group.add_argument('--file', action="store", default=None)

    arguments = parser.parse_args()

    if arguments.mode=="init":
        initialize_project_directory(Path('.'))
    elif arguments.mode=="compile":
        compile_all_files(Path('contracts'), Path('build') / Path('contracts'), Path("migrations"))
    elif arguments.mode=="server":
        run_server(Path("datadir"), arguments.httpserver, arguments.websocket)
    elif arguments.mode=="keyfile":
        if arguments.keyfile_mode == "encrypt":
            Authentication.encrypt_keyfile(arguments.keyfile_file, arguments.keyfile_private_key, arguments.keyfile_password)
        elif arguments.keyfile_mode == "decrypt":
            Authentication.decrypt_keyfile(arguments.keyfile_file, arguments.keyfile_password)
    elif arguments.mode=="epm":
        epm = PackageManager(Path(ETHEREUM_PACKAGES_DIR))
        epm.operate(arguments.epm_mode, arguments.epm_uri, arguments.epm_package)
    elif arguments.mode=="generate_mnemonic":
        mnemonic = Mnemonic.generate_mnemonic(int(arguments.num_words), arguments.language)
        print(mnemonic)
    elif arguments.mode=="test":
        pytest_args = arguments.file.split(" ")
        run(["py.test", *pytest_args])


if __name__ == "__main__":
    parse_cli_args()
